sauron/engine.py

```diff
@@ -44,43 +49,59 @@ def my_function():
             return None
         """
 
-        def decorator(function):
-            verbose_name = kwargs.get("verbose_name", None)
+        def decorator(function: Callable):
+            verbose_name: str = kwargs.get("verbose_name", None)
             if args:
                 verbose_name = args[0]
             self.__add_callable(function, verbose_name)
             return function
 
         return decorator
 
-    def apply_job_call(self, func_name, session):
-        target_func = self.callables_collected[func_name]["function"]
-        job_args = [
-            job.args for job in self.parsed_jobs.jobs if job.name == func_name
-        ]
-        if job_args:
-            job_args = job_args[0]
+    def apply_job_call(
+        self, job: JobModel, session: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        target_func: Callable = self.callables_collected.get(job.name).get(
+            "function"
+        )
+        if job.args:
+            result = target_func(session=session, **job.args)
         else:
-            raise ValueError(f'No job added with function name: {func_name}')
+            result = target_func(session=session)
+        # append result of function called into session
+        results = session.get("results", None)
+        if not results:
+            session["results"] = []
+        session["results"].append({"job": job.name, "return": result})
+        self.session = session
+        return session
+
+    def parse(self, unparsed_rule: Union[str, Dict[str, Any]]):
+        """
+            Parses rules
+        """
+        parser: DefaultParser = self.parser_class()
+        parsed_rule: List[JobModel] = parser.parse(unparsed_rule)
+        self.parsed_rule = parsed_rule
+        return parsed_rule
+
+    def run(
+        self, rule: Union[str, Dict[str, Any]], session: Dict[str, Any] = None
+    ):
+        """
+        Executes each job passing the current session to them
+        """
 
-        if job_args:
-            func_result = target_func(session=session, **job_args)
+        if session:
+            session = session
         else:
-            func_result = target_func(session=session)
-
-        # append result of function called into sessino
-        session[func_name] = {"return": func_result}
+            session = self.session
 
+        for job in self.parse(rule):
+            print(job)
+            print(session)
+            session = self.apply_job_call(job, session)
 
-    def parse(self, unparsed_jobs):
-        p = self.parser_class()
-        parsed_jobs = p.parse(unparsed_jobs)
-        self.parsed_jobs = parsed_jobs
-
-    def run(self, jobs, session):
-        """
-        Code that executes each Job respecting the jobs sequence
-        """
-        self.parse(jobs)
-        for job in self.parsed_jobs.jobs:
-            self.apply_job_call(job.name, session)
+    def export_metadata(self, fmt: str = "dict"):
+        exporter = DefaultExporter()
```

Comment:
the exporter should be an attribute of the class, so that way could be customized when creating a new instance, or by default will be a `DefaultExporter` instance

Reply:
meaning, 
```python
class Engine:
    def __init__(.... exporter_class=DefaultExporter):
        # constructor can receive which exporter (class or callable) will be used, otherwise will be the default one
        self.exporter_class = exporter_class

    ...

    def export_metadata(self, fmt: str = "dict"):
        p = self.parser_class()
        exporter_instance = self.exporter_class(...)
        return exporter_instance.export_jobs(self.callables_collected, fmt=fmt)
```

Reply:
Nice. This puts dependency injection to good use